### ISSUE TITLE
How can I get the size of the backups and snapshots ?

### ISSUE THREAD

**Issue by Lyndon-Li** [vmware-tanzu/velero#6506]
### Discussed in https://github.com/vmware-tanzu/velero/discussions/5466

<div type='discussions-op-text'>

<sup>Originally posted by **camillebernier** October 19, 2022</sup>
Hello,
I'm using Velero with Restic (for the volume snapshots) and Minio. 
All installed with Helm on Kubernetes.

Is there any way to get the size of the Backups and Snapshots (done with Restic)?
And to have details on each backup for example: 
- Size when the backup is incremental
- Size for a full backup
- Size for a snapshot
- Size before and after compression 
I can't find a way to do this, except by making a backup of a single namespace and its resources, and check the size of my bucket after.

Thanks</div>

**Comment by reasonerjt** [comment]
There's not an easy way to get the repo's size of restic/kopia, maybe we can enhance the backup repo interface with some code change in restic/kopia to get such information.

Let's keep this in the backlog.

**Comment by djjudas21** [comment]
Killer feature for me. At the moment I have to go and look in the S3 storage to see how much bigger it is than the previous day. Would be great to see how big the deltas are with the `velero` command.

**Comment by reasonerjt** [comment]
@shubham-pampattiwar 

As I mentioned in the earlier comment it is difficult to implement, do you have any concrete ideas?

**Comment by shubham-pampattiwar** [comment]
Velero tracks size information for most backup components, but this information is not currently exposed in the CLI or via Backup CR status.

  **Currently Available Size Information**

  -   **CSI Snapshots**: Size is stored in the [`CSISnapshotInfo`](https://github.com/vmware-tanzu/velero/blob/main/internal/volume/volumes_information.go#L136)
  struct.
      ```go
      type CSISnapshotInfo struct {
          // The snapshot corresponding volume size.
          Size int64 `json:"size"`
      }
      ```
      This is populated from `volumeSnapshot.Status.RestoreSize` at [line
  463](https://github.com/vmware-tanzu/velero/blob/main/internal/volume/volumes_information.go#L463).
      ```go
      CSISnapshotInfo: &CSISnapshotInfo{
          VSCName:        *volumeSnapshot.Status.BoundVolumeSnapshotContentName,
          Size:           size, // from volumeSnapshot.Status.RestoreSize.Value()
          Driver:         volumeSnapshotContent.Spec.Driver,
      }
      ```

  -   **File System Backups (Kopia)**: Size is stored in the
  [`PodVolumeInfo`](https://github.com/vmware-tanzu/velero/blob/main/internal/volume/volumes_information.go#L218) struct.
      ```go
      type PodVolumeInfo struct {
          // The snapshot corresponding volume size.
          Size int64 `json:"size,omitempty"`
      }
      ```
      This is populated from `pvb.Status.Progress.TotalBytes` at [line
  244](https://github.com/vmware-tanzu/velero/blob/main/internal/volume/volumes_information.go#L244).
      ```go
      func newPodVolumeInfoFromPVB(pvb *velerov1api.PodVolumeBackup) *PodVolumeInfo {
          return &PodVolumeInfo{
              SnapshotHandle: pvb.Status.SnapshotID,
              Size:           pvb.Status.Progress.TotalBytes, // from Kopia
              UploaderType:   pvb.Spec.UploaderType,
          }
      }
      ```

  -   **CSI Data Mover (Kopia)**: Size is stored in the
  [`SnapshotDataMovementInfo`](https://github.com/vmware-tanzu/velero/blob/main/internal/volume/volumes_information.go#L171) struct.
      ```go
      type SnapshotDataMovementInfo struct {
          // Moved snapshot data size.
          Size int64 `json:"size"`
      }
      ```
      This is populated from `dataUpload.Status.Progress.TotalBytes` at [line 1200](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/backup.go#L1200).
      ```go
      volumeInfos[index].SnapshotDataMovementInfo.Size = dataUpload.Status.Progress.TotalBytes
      ```

  -   **K8s Objects**: The tarball size is calculated at [line 833](https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/backup_controller.go#L833) but
  currently only used for metrics.
      ```go
      var backupSizeBytes int64
      if backupFileStat, err := backupFile.Stat(); err != nil {
          log.WithError(errors.WithStack(err)).Error("Error getting backup file info")
      } else {
          backupSizeBytes = backupFileStat.Size()
      }
      serverMetrics.SetBackupTarballSizeBytesGauge(backupScheduleName, backupSizeBytes)
      ```

  **Missing Size Information**

  -   **Native Cloud Provider Snapshots**: There is no size field for native snapshots. This would require cloud provider API calls to implement. I think we can put this on the back burner for now.

  **Proposed Solution**

  **Phase 1: Add Size Fields to `BackupStatus`**

  Extend [`BackupStatus`](https://github.com/vmware-tanzu/velero/blob/main/pkg/apis/velero/v1/backup_types.go) in `pkg/apis/velero/v1/backup_types.go`:

  ```go
  type BackupStatus struct {
      // ... existing fields ...

      // BackupSizeBytes is the size of the K8s objects backup tarball
      BackupSizeBytes *int64 `json:"backupSizeBytes,omitempty"`

      // VolumeSnapshotSizeBytes is the total size of all volume snapshots
      VolumeSnapshotSizeBytes *int64 `json:"volumeSnapshotSizeBytes,omitempty"`

      // Optional: Breakdown by backup method
      VolumeSnapshotSizeBreakdown *VolumeSnapshotSizeBreakdown `json:"volumeSnapshotSizeBreakdown,omitempty"`
  }
```

  Phase 2: Aggregate and Display

  - Modify the https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/backup_controller.go to aggregate sizes from the existing
  https://github.com/vmware-tanzu/velero/blob/main/internal/volume/volumes_information.go#L55 data.
  - Update https://github.com/vmware-tanzu/velero/blob/main/pkg/cmd/util/output/backup_describer.go to display the size information in velero backup describe.
```yaml
  Example output:

  Phase:                    Completed
  Backup Size:              1.2 GB (K8s objects)
  Volume Snapshot Size:     850 MB (total)
    - CSI Snapshots:        340 MB (2 volumes)
    - File System Backups:  385 MB (1 volume)
    - Data Movement:        125 MB (1 volume)
  ```
@Lyndon-Li @reasonerjt This is just a high level overview, Let me know you guys think.

**Comment by Lyndon-Li** [comment]
@shubham-pampattiwar 
I think the current issue is not only for the existing logical size, but also for physical size in the backup repo, incremental size for each incremental backup, etc.

**Comment by sseago** [comment]
Right. And some things contribute to one but not the other. CSI and native snapshots don't take up any space in the BSL, for example. BUt you also have log sizes -- if we're talking about BSL usage, then backup logs take up room just as k8s metadata does. I'm not sure how we'd track or report incremental size. For one thing, getting that information out of kopia may be difficult. Also, it will change over time when prior backups expire. For example, if you have backup1 with a single volume, then take backup2 of the same volume, then the size taken up by backup2 in s3 will be smaller than backup1 due to the incremental nature of it, but once backup1 expires, then the prior full volume backup that backup2's incremental backup use is now only associated with backup2 -- if we don't start reporting that with backup2, then the usage totals won't match.

**Comment by weshayutin** [comment]
@sseago @reasonerjt @shubham-pampattiwar would it be possible to break this issue up into parts?
* part-1: everyone agrees on $thing and it's technically feasible in v1.18
* part-2: everyone agrees on $thing and it may take more time than one release
* part-3: items where agreement has not yet been found?

**Comment by weshayutin** [comment]
@sseago I belive this is the issue you have been working to some degree in the background.  Any updates here, can we get it on the milestone for 1.18 or partially?

**Comment by sseago** [comment]
@weshayutin Sure. The thing I'm working on is just a part of this -- reporting on incremental size for kopia filesystem backup and datamover. I guess I should create a separate issue to track just that portion. The larger task of "how big is this backup?" still needs further definition.

**Comment by Lyndon-Li** [comment]
A partial fix for `incrementalSize` would be in 1.18, for others, let's continue the work post 1.18.

**Comment by weshayutin** [comment]
@Lyndon-Li @sseago @shubham-pampattiwar how aboooout 1.20-candidate :) \0 0/ 0_ 0| \0

**Comment by Lyndon-Li** [comment]
We have supported both the full size and incremental size for fs data mover/fs-backup, block data mover will also support this.
For data storage size, let me further think about.
